nixos/postgresql: move postStart into separate unit

wolfgangwalther opened the pull request:
This avoids restarting the postgresql server, when only `ensureDatabases` or `ensureUsers` have been changed. Later, it will also allow to properly wait for recovery to finish (https://github.com/NixOS/nixpkgs/issues/346886#issuecomment-2839062089).

To wait for "postgresql is ready" in other services, we now provide a `postgresql.target`.

Supersedes #400018

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [x] (Module updates) Added a release notes entry if the change is significant
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

wolfgangwalther committed:
nixos/postgresql: align maintainers with postgresql package

wolfgangwalther committed:
nixos/postgresql: move postStart into separate unit

This avoids restarting the postgresql server, when only ensureDatabases
or ensureUsers have been changed. It will also allow to properly wait
for recovery to finish later.

To wait for "postgresql is ready" in other services, we now provide a
postgresql.target.

Resolves #400018

Co-authored-by: Marcel <me@m4rc3l.de>

Ma27 reviewed (commented):
That's a meh from me: I'm not a fan at all to make the module more complex for a broken feature. Not only because it's just poorly implemented convergence in a trenchcoat, also given the issues we've seen in the past (me and Ryan having to fix up numerous modules because the old implementation broke with Postgresql 15 right before the 23.11 release or this feature being outright broken after a glibc upgrade).

This change is in terms of complexity small enough to just do it, so I won't actively block it, however I also don't plan to fix any issues in the module associated with ensure*-style options, the -setup.service or support people encountering issues with that.

I should mention that I'm not only complaining, but as part of my Bachelor's thesis I do evaluate an approach on how to do state-management in a more principled way, i.e. without ad-hoc bash scripts wherever needed. I'll first need to get this thesis stuff done, but I'll hopefully get around to publish a proposal this Summer. So I'm kinda hopeful that we can talk about potential alternatives for this later this year ;)

Ma27 commented:
> WDYT @Ma27? I don't want to add stuff that you're explicitly not supportive of - but at the same time we need to solve the "recover from backup" story in a sensible way.

Sorry, I didn't reply because of disagreement, but because of finishing thesis things (that's the main reason why reviews are slower than they were a few months ago, your work is nonetheless still appreciated!). If I could get ~another day to reply, that'd be highly appreciated.

wolfgangwalther committed:
nixos/postgresql: make postgresql.target wait until recovery is done

The new postgresql.target will now wait until recovery is done and
read/write connections are possible.

This allows ensure* scripts and downstream migrations to work properly
after recovery from backup.

Resolves #346886

Ma27 commented:
> Given that this is a change that will potentially require downstream users to adjust, I'm wondering whether we should wait for branch-off or not. On the one side this reduces risk, because there is not much time to test it outside nixpkgs between now and the release. On the other side, this would mean that recovery would stay broken in many cases for the 25.05 cycle.

Hmm now that you say it: I think we've accumulated quite a few changes already and we're pretty close to branch-off, so it may make sense to delay this to 25.11.

The issue is absolutely valid, but I don't see a sufficiently strong demand to fast-track this now. Also, I'd expect most setups exceeding a given size to have a dedicated DB host, so the problem of other units being started too early is not too common there.

So yeah, the release-notes bullet-point needs to wait until branch-off ;-)

Ma27 commented on nixos/modules/services/databases/postgresql.nix:
```diff
@@ -718,12 +718,23 @@ in
       after = [ "network.target" ];
 
+      # To trigger the .target also on "systemctl start postgresql".
+      bindsTo = [ "postgresql.target" ];
```
I must confess, I tried to be smarter than I am on this.
BindsTo only propagates its action (e.g. start/restart/stop) to the listed units within the same transaction.
However, it does not impose any ordering requirements.

It's just that I've seen the Requires/BindsTo+After combination too often that I snapped on this. Sorry :see_no_evil:

Ma27 commented on nixos/modules/services/databases/postgresql.nix:
```diff
@@ -850,7 +850,7 @@ in
       script =
         ''
-          while ! psql -d postgres -c "" 2> /dev/null; do
+          while ! psql -d postgres -c 'DO $$ BEGIN IF pg_is_in_recovery() THEN RAISE read_only_sql_transaction; END IF; END$$;' 2> /dev/null; do
```
Not the nicest approach, but what about a shell functions that returns non-zero if either the psql `SELECT pg_is_in_recover()` invocation fails due to a connection error) or if the returned value is `0`?

Ma27 commented on nixos/modules/services/databases/postgresql.nix:
```diff
@@ -862,11 +830,66 @@ in
+      script =
+        ''
+          while ! psql -d postgres -c 'DO $$ BEGIN IF pg_is_in_recovery() THEN RAISE read_only_sql_transaction; END IF; END$$;' 2> /dev/null; do
+              if ! systemctl is-active --quiet postgresql.service; then exit 1; fi
```
Doesn't this need a `config.systemd.package` in `path`?

I'm wondering if it'd make sense to actually wait for something where we can reasonably test this, even if this means waiting for pgbackrest[1]. I read over this three times before noticing it.

[1] damn it, that means we have a cyclic dependency, right?

Ma27 replied:
Ah OK. My reading was that pgbackrest is a way to test the behavior in here properly, but that this fix would also be necessary for the tests to pass. Fine then. Leaving unresolved given the point about test coverage in this PR still stands.

Ma27 commented on nixos/modules/services/databases/postgresql.nix:
```diff
@@ -872,11 +840,74 @@ in
+            EOF
+          }
+          while ! check-connection 2> /dev/null; do
+              if ! systemctl is-active --quiet postgresql.service; then exit 1; fi
```
> not at all, actually..., but it doesn't really need to be

AFAICS applying the same hardening as for postgresql.service seems to work for `systemctl is-active` (and given that AF_UNIX is allowed, I'm pretty sure also for the psql part). So maybe try it out?

Given its low surface, I wouldn't recommend spending too much time on it, though.

Ma27 replied:
My personal stance on this is far simpler: the code-path doesn't need certain privileges, so it shouldn't have those. So my approach would've been to just re-use the postgresql.service hardening and call it a day.

Anyways, this is not critical and can be done in a follow-up. So, I'll close this now.

Ma27 reviewed (changes requested):
The implementation looks good and I agree with the approach.

The following things are IMHO tbd to merge:

* check if we can trivially apply a bunch of hardening settings to postgresql-setup
* add a note to the release notes
* replace target in new modules (a quick grep gives me lasuite-docs & mealie).

Thanks for your work here!

Ma27 commented on nixos/doc/manual/release-notes/rl-2511.section.md:
```diff
@@ -42,6 +42,8 @@
 
 - The Pocket ID module ([`services.pocket-id`][#opt-services.pocket-id.enable]) and package (`pocket-id`) has been updated to 1.0.0. Some environment variables have been changed or removed, see the [migration guide](https://pocket-id.org/docs/setup/migrate-to-v1/).
 
+- The `services.postgresql` module now sets up a systemd unit `postgresql.target`, which other services should depend on instead of `postgresql.service`. Only `postgresql.target` will have ensured users and run the initial setup script. Additionally, it will wait for a read-write connection after a restore from backup, while `postgresql.service` will only wait for a read-only connection during re ...
```
Perhaps something along the lines of

> Depending on `postgresql.target` guarantees that postgres is in read-write mode and initial/ensure scripts were executed. Depending on `postgresql.service` only guarantees a read-only connection

Ma27 reviewed (approved):
Looks good now, thank you!

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/breaking-changes-announcement-for-unstable/17574/84
